_blogposts/compiler/string-literal-types-in-reason.mdx

```diff
@@ -0,0 +1,215 @@
+JavaScript behavior, it's a relatively new concept since most type systems are runtime encoding agnostic. However, to smooth the user experience when writing
+bindings to existing JS API, we are introducing string literal types which is unique in several behaviors compared with TypeScript: it supports type inference,
+pattern match and can be attached to data.
+
+## Vanilla string literal types
+
+The notation in Reason for string literal types is like this \``hello`, it will be compiled into "hello", the difference is that \``hello` is given a type so that you can not mix it with other strings.
+
+Take such code snippet below as an example:
+
+```reasonml
+let encoding = (enc) =>
+  switch (enc) {
+  | `utf8 => 0
+  | `ascii => 1
+  | `utf16 => 2
+  };
+```
+
+It will be compiled into
+
+```js
+function encoding(en) {
+  if (en === "ascii") {
+    return 1;
+  } else if (en === "utf16") {
+    return 2;
+  } else {
+    return 0;
+  }
+}
+```
+
+If you pass a random encoding, e.g, `` encoding (`ucs32) ``, you get a type error:
+
+```
+This expression has type [> `ucs32 ]
+       but an expression was expected of type [< `ascii | `utf16 | `utf8 ]
+       The second variant type does not allow tag(s) `ucs32
+```
+
+Another thing you can observe from the generated JS is that since the compiler can guarantee that the input could only be `` `utf8, `ascii,`utf16 ``, it will skip the comparison with "utf8" when the first two are compared.
+
+If we add a wild catch to match any encoding
+
+```reasonml
+let encoding = (enc) =>
+  switch (enc) {
+  | `utf8 => 0
+  | `ascii => 1
+  | `utf16 => 2
+  | _ => 3
+  };
+```
+
+It will generate JS as below:
+
+```js
+function encoding(en) {
+  if (en === "utf8") {
+    return 0;
+  } else if (en === "ascii") {
+    return 1;
+  } else if (en === "utf16") {
+    return 2;
+  } else {
+    return 3;
+  }
+}
+```
+
+## Declaring types for string literal types
+
+Note all string literal types can be inferred, this is great for develop experience when you are doing the development. When things get more stable, it would be nice to give string literal types a name as below:
+
+```reasonml
+type utf = [
+  | `utf8
+  | `utf19
+];
+type ascii = [
+  | `ascii
+]
+
+```
+
+The cool thing is that you can creat union types by simply put them together:
+
+```reasonml
+type encoding = [
+  | utf
+  | ascii
+]
+```
+
+The compiler even supports a sugar over named string literal types:
+
+```reasonml
+let classify = (enc) =>
+  switch (enc) {
+  | #utf => "utf" // string literals belong to utf
+  | #ascii => "asci" // string literals belog to ascii
+  };
+```
+
+The compiler would generate well optimized code as below:
+
+```js
+function classify(enc) {
+  if (enc === "ascii") {
+    return "asci";
+  } else {
+    return "utf";
+  }
+}
+```
+
+## String literal types in bindings
+
+Since string literal types are just strings after type checking, user can use it
+bind to js libs directly without any conversion as below:
+
```

Comment:
Since string literal types are just strings after type checking, you can use them to bind to JS libraries directly without any conversion, as follows: